Fix key value metadata

ap-- opened the pull request:
Closes #751

Hello everyone,

I added a few tests to define the behavior of the key_value_metadata.
It's now rejecting non-bytes and non-str keys and values.
And bytes always get decoded if possible.

Cheers,
Andreas

martindurant commented on fastparquet/api.py:
```diff
@@ -261,8 +261,18 @@ def statistics(self):
+                try:
+                    key = kv.key.decode()
+                except UnicodeDecodeError:
+                    key = kv.key
```
I wonder whether it should simply be rejected (same below), omitted or error. Can you try if other parquet frameworks (arrow, spark, older fastparquet) can recover non-utf8 bytestrings?

yohplala commented:
> @yohplala , appreciate your thoughts on this one.

Ok, I will have a look tomorrow.
I need to run your code @ap-- that you provided last in #751.

I am "problem-solving" oriented, and I don't understand the problem.
Hence I don't understand why we need all these checks everywhere:

```python
 if isinstance(value, str) else value
```

Obviously, it should not hurt performance, but then currently I don't understand what they solve.
Does it solve an existing trouble or are we imagining there could be a trouble and in anticipation we take measures to prevent it? if this is the latter, then, following this logic, in the future, the code may get very "heavy" by adding new checks "in case" (and I would add, it is already complex enough - some people already told so in their tickets in the past)

I did not take time to understand yet the points you raise in your [last comment](https://github.com/dask/fastparquet/issues/751#issuecomment-1222544378) in #751.
Definitely, I take a look tomorrow.

martindurant commented:
I think the basic "problem"  is that parquet stores utf8 bytestrings, so we need to enforce what the user might provide at the time that we save. Otherwise, you can get the sorts of problems in the linked issue.

yohplala commented:
Hi @martindurant , hi @ap-- ,

Ok sorry, I was not in the topic so far.
I ran your code @ap-- with last source and got the same error message.

```python
AttributeError: 'int' object has no attribute 'decode'
```

Ok, here are my feedbacks:

- Should `fastparquet` accept only `str` for metadata or both `str` and bytestrings for cutom metadata?
  This, I am not really able to say what is adequate. I currently stick to the first (only `str`) but then if other usages require also bytestrings, why not. I believe it adds checks and I am not so fond of that. But then this should certainly not block the initiative if this is needed.

- **Code simplification**.
  @ap-- , I am sorry, I have not tested, but I believe the type check could be commonalized between it being in `writer.write()` and `util.update_custom_metadata()`

The check I am speaking about:
```python
        if not isinstance(key, (bytes, str)):
            raise ValueError(f"custom_metadata key not a string: {key!r}")
        if not (value is None or isinstance(value, (bytes, str))):
            raise ValueError(f"custom_metadata value not a string: {value!r}")
```

- in `writer.write_simple()`, you could replace line
`        foot_size = f.write(fmd.to_bytes())`
with
`        foot_size = write_thrift(f, fmd)`

Then you can position the check in `writer.write_thrift()` that actually reads:
```python
def write_thrift(f, obj):
    # TODO inline this
    return f.write(obj.to_bytes())
```

So the new one would then read (something like - again, untested):
```python
def write_thrift(f, obj):
    for kv in fmd.key_value_metadata:
        if not isinstance(kv.key, (bytes, str)):
            raise ValueError(f"custom_metadata key not a string: {kv.key!r}")
        if not (kv.value is None or isinstance(value, (bytes, str))):
            raise ValueError(f"custom_metadata value not a string: {kv.value!r}")
    # TODO inline this
    return f.write(obj.to_bytes())
```
@martindurant would you be ok with that?
@ap-- do you think you can give it a try?

ap-- commented:
Yes, basically it all boils down to three points:

1. what types of keys and values should fastparquet allow to be provided to custom_metadata on write? (and I think the answer here is str and bytes)
2. once (1) is restricted in a newer version of fastparquet: what should fastparquet do with non-standard parquet files that can be created with current versions by providing for example `int` type values in custom metadata
3. in `ParquetFile.key_value_metadata` should keys and values always be returned as str? and if yes, what if they are non-utf8


I checked the behavior of `pyarrow`. See the pytest file in the code block below. (Let me know if I should add it to the PR)


```python
import pytest
import pandas as pd
from fastparquet import write
from pyarrow import Table
from pyarrow.parquet import read_table


@pytest.fixture
def df():
    yield pd.DataFrame(data={'a': [1]})


def _kv_id(x):
    try:
        x.decode()
    except UnicodeDecodeError:
        return "non_utf8"
    except AttributeError:
        pass
    return type(x).__name__


@pytest.mark.parametrize(
    "key,value", [
        ("k", "v"),
        ("k", b"v"),
        (b"k", "v"),
        (b"k", b"v"),
        (b"\xe2", b"v"),
        (b"k", b"\xe2"),
    ],
    ids=_kv_id
)
def test_pyarrow_metadata_bytes_str_handling(df, key, value):
    tbl = Table.from_pandas(df)
    md = tbl.schema.metadata or {}
    md[key] = value
    tbl2 = tbl.replace_schema_metadata(md)
    md2 = tbl2.schema.metadata

    md2.pop(b"pandas")  # easier to read pytest output
    assert {key: value} == md2


@pytest.mark.parametrize(
    "value", [1, 1.0, 1j, True, None, (), [], {}],
    ids=_kv_id
)
def test_pyarrow_reject_non_str_bytes_value(df, value):
    tbl = Table.from_pandas(df)
    md = tbl.schema.metadata or {}
    md["k"] = value

    with pytest.raises(TypeError):
        _ = tbl.replace_schema_metadata(md)


@pytest.fixture
def non_std_md_parquet_fn(request, df, tmp_path):
    value = request.param
    if isinstance(value, (complex, tuple)):
        pytest.skip(f"{type(value).__name__} values segfault fastparquet")
    fn = str(tmp_path.joinpath("non_std_md.parquet"))
    write(fn, df, custom_metadata={"non_std_key": value})
    yield fn


@pytest.mark.parametrize(
    "non_std_md_parquet_fn",
    [1, 1.0, 1j, True, None, (), [], {}],
    ids=_kv_id,
    indirect=True,
)
def test_pyarrow_parsing_non_str_kv(non_std_md_parquet_fn):
    tbl = read_table(non_std_md_parquet_fn)
    md = tbl.schema.metadata

    md.pop(b"pandas")  # easier to read pytest output
    assert {"non_std_key": 1} == md

```

To summarize `pyarrow` behavior:

1. pyarrow rejects any key or value that is not a str or bytes instance
2. pyarrow returns `b""` instead of non str/bytes values in the metadata
3. pyarrow always returns bytes, and doesn't try to decode


Cheers,
Andreas

P.S.: This also uncovered another bug currently in fastparquet: either of the two lines below segfault fastparquet. (complex and tuple values)
```python
fastparquet.write("out.parq", df, custom_metadata={"k": 1j})
fastparquet.write("out.parq", df, custom_metadata={"k": ()})
```

yohplala commented:
Thanks @ap-- for this concise summary:

>Yes, basically it all boils down to three points:
> 1- what types of keys and values should fastparquet allow to be provided to custom_metadata on write? (and I think the answer here is str and bytes)
> 2- once (1) is restricted in a newer version of fastparquet: what should fastparquet do with non-standard parquet files that can be created with current versions by providing for example int type values in custom metadata
> 3- in ParquetFile.key_value_metadata should keys and values always be returned as str? and if yes, what if they are non-utf8

1- Hm, ok for me to allow `str` and bytestring then.

2- Hmm, I would say the breaking change already occured between 0.8.0 (no error) and 0.8.1 (now an error). How fastparquet currently behave with that? A file created with 0.8.0, with an `int`, and now read with 0.8.1? If there is no error, I would leave it as it is.

3- Hmmm, logic I had so far is that user would only record `str` metadata so, it makes sense in this case to return him/her `str` only metadata. But now, with (1), this logic does not apply any longer.
 Do the tests you make in `api.ParquetFile.key_value_metadata()` to check if they can be decoded or not make it possible to manage this case of non-utf8 `str`?
```python
            for kv in self.fmd.key_value_metadata or []:
                try:
                    key = kv.key.decode()
                except UnicodeDecodeError:
                    key = kv.key
                try:
                    value = kv.value.decode()
                except (UnicodeDecodeError, AttributeError):
                    value = kv.value 
```
?
If yes, so you also brought the solution, right?
- if ut8 `str`, then they are decoded,
- if they are not, they are not decoded.

ap-- commented:
> Thanks @ap-- for this concise summary:
>
> [...]
> 
> 1- Hm, ok for me to allow `str` and bytestring then.

Good 👍 
 
> 2- Hmm, I would say the breaking change already occured between 0.8.0 (no error) and 0.8.1 (now an error). How fastparquet currently behave with that? A file created with 0.8.0, with an `int`, and now read with 0.8.1? If there is no error, I would leave it as it is.

"non-standard-custom-metadata" parquet files (i.e. `int` values) can be written with fastparquet `0.8.0+` and newer.

I also tested `0.7.x` just now and those versions don't allow writing it, and raise a `TypeError` like `pyarrow`. Since `0.8.1+` those "non-standard-custom-metadata" parquet files can't be read anymore. 

A "non-standard-custom-metadata" parquet file written with `0.8.x` and read with `0.7.x` raises an Exception:

```python
>>> fastparquet.ParquetFile("./temp0.8.0.parquet")
Traceback (most recent call last):
  File "/Users/poehlmann/Development/_fp_test/myvenv/lib/python3.8/site-packages/fastparquet/api.py", line 187, in _parse_header
    fmd = read_thrift(f, parquet_thrift.FileMetaData)
  File "/Users/poehlmann/Development/_fp_test/myvenv/lib/python3.8/site-packages/fastparquet/thrift_structures.py", line 25, in read_thrift
    obj.read(pin)
  File "/Users/poehlmann/Development/_fp_test/myvenv/lib/python3.8/site-packages/fastparquet/parquet_thrift/parquet/ttypes.py", line 4446, in read
    iprot._fast_decode(self, iprot, [self.__class__, self.thrift_spec])
TypeError: got wrong ttype while reading field

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/poehlmann/Development/_fp_test/myvenv/lib/python3.8/site-packages/fastparquet/api.py", line 133, in __init__
    self._parse_header(f, verify)
  File "/Users/poehlmann/Development/_fp_test/myvenv/lib/python3.8/site-packages/fastparquet/api.py", line 189, in _parse_header
    raise ParquetException('Metadata parse failed: %s' % self.fn)
fastparquet.util.ParquetException: Metadata parse failed: /Users/poehlmann/Development/_fp_test/./temp0.8.0.parquet
```


> 3- Hmmm, logic I had so far is that user would only record `str` metadata so, it makes sense in this case to return him/her `str` only metadata. But now, with (1), this logic does not apply any longer. Do the tests you make in `api.ParquetFile.key_value_metadata()` to check if they can be decoded or not make it possible to manage this case of non-utf8 `str`?

In the tests I use `b"\xe2"` as an example bytestring that can't be utf8 decoded.

> ? If yes, so you also brought the solution, right?

Yes. I'm sorry if I didn't communicate this clearly before.


<br><br>

Given the testing with `pyarrow` and my comparisons with `fastparquet==0.7.x` I think there would be two minor things to consider:
- we should raise a `TypeError` when custom_metadata keys or values are not str/bytes, instead of a ValueError in the current PR
- by catching the AttributeError in `api.ParquetFile.key_value_metadata` I currently return the actual `int` value for those non-standard custom_metadata parquet files written with fastparquet. pyarrow replaces those values with `b""`. Should fastparquet do the same (and maybe emit a warning) or should it raise an Exception like `fastparquet==0.7.x`?

yohplala commented:
> A "non-standard-custom-metadata" parquet file written with 0.8.x and read with 0.7.x raises an Exception.

I am more interested in the opposite: writing with 0.7.x (so some 'old' data a user may have in his/her datawarehouse) and reading with 0.8.x.
For the scenario you state, the answer would be, I guess: "Please update your fastparquet version".

> we should raise a TypeError when custom_metadata keys or values are not str/bytes, instead of a ValueError in the current PR

I think you are right, it makes more sense.

> by catching the AttributeError in api.ParquetFile.key_value_metadata I currently return the actual int value for those non-standard custom_metadata parquet files written with fastparquet. pyarrow replaces those values with b"". Should fastparquet do the same (and maybe emit a warning) or should it raise an Exception like fastparquet==0.7.x?

I think the solution you propose is good. It has 2 aspects:
- returning a value:
   well, why not. It might be better than returning nothing like pyarrow.
- not raising an exception like fastparquet 0.7.x:
   it allows the user to retrieve his/her dataframe, even if the metadata are not retrieved, which is  really better than nothing.

martindurant commented:
OK, I agree with the whole conversation here:
- we allow str and bytes in metadata
- on load, we attempt to decode bytes to str, but leave as bytes on failure (possible alternative here is to decode with `errors="replace"` or `errors="ignore"`)
- on load, if the key/value is anything other than bytes or str, this is not strictly valid, and we pass it through unchanged.

ap-- commented:
Sounds good! I'll implement the changes.

ap-- commented:
I refactored the tests, added an example file with non-std metadata with a test, and refactored the fix with your comments in mind @yohplala.

This is ready for another review :+1:

yohplala commented:
> I refactored the tests, added an example file with non-std metadata with a test, and refactored the fix with your comments in mind @yohplala.
> 
> This is ready for another review +1

Hi @ap-- , this seems to me very good work, thank you for the quality of this PR and your patience for it!

martindurant commented:
Thank you. This is going in, and has made the bugfix release I am about to put out.
